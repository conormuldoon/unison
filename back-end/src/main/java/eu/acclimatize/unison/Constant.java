package eu.acclimatize.unison;

public class Constant {
	
	public static final String LOCATION = "location";
	public static final String FORMAT = "d/M/yyyy";
	public static final String FROM_DATE = "fromDate";
	public static final String TO_DATE = "toDate";
	
	public static final String USERNAME = "username";
	public static final String PASSWORD = "password";
	
	public static final String PROPERTIES_FILE = "/application.properties";
	
	


}
