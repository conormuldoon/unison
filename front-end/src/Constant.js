
export const FORMAT = "D/M/YYYY";

export const API = "/apiacc";

export const LPER = "20\u1d57\u02b0 percentile";
export const HPER = "80\u1d57\u02b0 percentile";
export const MEDIAN = "Median";

export const LOW = "Low";
export const MEDIUM = "Medium";
export const HIGH = "High";
export const PRECIP = "Precipitation";
export const CL = "Cloud Level";
//export const API = "";

// Map coordinates
//export const CENTRE_LAT = 53.35014;
//export const CENTRE_LON = -6.266155;

// Oslo
export const CENTRE_LAT = 59.922326;
export const CENTRE_LON = 10.751560;
