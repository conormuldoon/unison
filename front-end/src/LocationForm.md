LocationForm example:

```js
import {useState} from 'react';
const [display,setDisplay] = useState(false);
function toggleDisplay(){
    if(display)setDisplay(false);
    else setDisplay(true);
}

<div style={{paddingLeft:200}}>
    <LocationForm display={display} toggleDisplay ={toggleDisplay} />
</div>
```