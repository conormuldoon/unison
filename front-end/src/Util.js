
import {FORMAT} from './Constant';

import  {
  formatDate
} from 'react-day-picker/moment';

export const fromDate = () =>{

  const today=new Date(Date.now());
  const fDate=new Date(Date.now());
  fDate.setMonth(today.getMonth()-1);

  return formatDate(fDate,FORMAT);
}

export const today = () => {
  return formatDate(new Date(Date.now()), FORMAT);
};

const postObject = (body) =>{
  return {
    method: 'POST',
    headers: new Headers({
               'Content-Type': 'application/x-www-form-urlencoded',
      }),
    body: body
  }
}

const locCred = (location,uname,pword) =>{
  return 'location='+location+'&username='+uname+'&password='+pword;
}

export const removePostObject = (location,uname,pword) =>{
  return postObject(locCred(location,uname,pword));

}

export const locationPostObject = (location,uname,pword,lon,lat) =>{
  return postObject(locCred(location,uname,pword)+'&longitude='+lon+'&latitude='+lat);
}

export const varMapping = (varCur) =>{
  let variRequest=varCur.replace(/ /g,'');
  variRequest=variRequest.charAt(0).toLowerCase() + variRequest.slice(1);
  return variRequest;
}
