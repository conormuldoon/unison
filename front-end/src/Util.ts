
import { FORMAT } from './Constant';

import {
  formatDate
} from 'react-day-picker/moment';

import GeoJSON from 'geojson';

/**
 * Cretes a string for tomorrow's date.
 * 
 * @returns Tomorrow's date formatted in terms of day of the month, month, and year.
 */
export const tomorrow = (): string => {

  const now = Date.now();
  const today = new Date(now);
  const tomorrow = new Date(now);
  tomorrow.setDate(today.getDate() + 1)

  return formatDate(tomorrow, FORMAT);
}

/**
 * Cretes a string for tomorros's date.
 * 
 * @returns Tomorrow's date formatted in terms of day of the month, month, and year.
 */
export const today = (): string => {
  return formatDate(new Date(Date.now()), FORMAT);
};

/**
 * Converts a weather variable that may be selected by the user to a mapping suffix for an endpoint
 * on the server. The initial character of the weather variable is changed to lower case and words 
 * are concatenated. For example, 'Wind Direction' is converted to 'windDirection'. 
 * 
 * @param varCur The weather variable.
 * @returns The converted string.
 */
export const varMapping = (varCur: string): string => {
  let variRequest = varCur.replace(/ /g, '');
  variRequest = variRequest.charAt(0).toLowerCase() + variRequest.slice(1);
  return variRequest;
}

/**
 * Displays a message to te user that there was a problem connecting to the server.
 */
export const problemConnecting = (): void => {
  alert('There was a problem connecting to Unison. Please try again later.');
}


/**
 * Changes second word in a weather variable to lower case if present so that the second
 * word is not capitalised in a sentence. For example, the weather variable 'Wind Direction' 
 * would be changed to 'Wind direction'.
 * 
 * @param curVar The weather variable to be changed.
 * @returns The converted string.
 */
export const chartText = (curVar: string): string => {

  const vca = curVar.split(' ');

  let vc = vca[0];

  if (vca.length > 1) {
    vca[1] = vca[1].toLowerCase();

    vc += ' ' + vca[1];
  }

  return vc;
}

/**
 * Obtains the CSRF token with the key XSRF-TOKEN from the stored cookie.
 * 
 * @returns The CSRF toke.
 */
export function csrfToken(): string {

  // Modified regular expression from an anwswer to https://stackoverflow.com/questions/51109559/get-cookie-with-react
  // and removed unnecessary slash.
  return document.cookie.replace(/(?:(?:^|.*;\s*)XSRF-TOKEN\s*=\s*([^;]*).*$)|^.*$/, "$1");

}


/**
 * Creates request options for making put requests to the server for JSON. 
 * 
 * @param body The JSON body of the request.
 * @returns The options created. The optinos include the a X-XSRF-TOKEN header for the CSRF token.
 */
const putObject = (body: string) => {
  return {
    method: 'PUT',
    headers: new Headers({
      'Content-Type': 'application/geo+json',
      'X-XSRF-TOKEN': csrfToken()
    }),
    body: body
  }

}

/**
 * Creates put options with a GeoJSON representation of the location in the body.
 * 
 * @param locationName The name of the location to be put to the server.
 * @param longitude The longitute coordinate of the location. 
 * @param latitude The latitude coordinate of the location. 
 * 
 * @returns The put options create, which include a X-XSRF-TOKEN header for the CSRF token.
 */
export const locationPutObject = (locationName: string, longitude: string, latitude: string):
  Record<string, unknown> => {

  const location = { name: locationName, lng: longitude, lat: latitude };

  const geoJSONPoint = GeoJSON.parse(location, { Point: ['lat', 'lng'] });
  return putObject(JSON.stringify(geoJSONPoint));
}
