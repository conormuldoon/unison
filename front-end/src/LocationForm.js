
import React, { useState } from 'react';
import {API} from './Constant';
import './App.css';
import {locationPostObject} from './Util';

import PropTypes from 'prop-types';


/**
 * A component for displaying a form to add new locations to be tracked.
 *  
 */
export default function LocationForm(props) {

  

  const [location,setLocation]=useState('');
  const [lon,setLon]=useState('');
  const [lat,setLat]=useState('');
  const [uname,setUName]=useState('');
  const [pword,setPWord]=useState('');


  function handleLoc(event){
    setLocation(event.target.value);

  }

  function handleLon(event){
    setLon(event.target.value);
  }

  function handleLat(event){
    setLat(event.target.value);

  }

  function handleUN(event){
    setUName(event.target.value);
  }

  function handlePassw(event){
    setPWord(event.target.value);
  }

  function handleSubmit(event){

    postData();

    event.preventDefault();

   }

   async function postData(){

    const response=await fetch(API+'/addLocation',
      locationPostObject(location,uname,pword,lon,lat));



    if(response.ok){
      const resVal=await response.json();

      if(resVal===0){
        alert('Location already exists');
      }else if(resVal===1){
        alert(location+' added');
        updateDisplay();

      }else if(resVal===2){
        alert('Incorrect user name or password');
      }else{
        alert(location + ' added, but did not obtain weather data.')
        updateDisplay();
      }
    }else{
      alert('Problem connecting');
    }

  }



  function updateDisplay(){
    setLocation('');
    setLon('');
    setLat('');
    setUName('');
    setPWord('');
    props.hideDisplay();
    props.obtainData();
  }


    let text='Add Location';
    if(props.display){
      text='Hide';
    }



    return ( <div id='pLeft'>

          <button data-testid="lf-button" onClick={props.toggleDisplay}>{text}</button>

          {props.display&&<form  id='locForm' onSubmit={handleSubmit}>
              <label>
                Location name: <input type="text" name="name" value={location} onChange={handleLoc} required  />
              </label>
              <br/>
              <br/>
              <label>
                Longitude: <input type="number" step="any" name="lon" value={lon} onChange={handleLon} required  />
              </label>
              <br/>
              <br/>
              <label>
                Latitude: <input type="number" step="any" name="lat"  value={lat} onChange={handleLat} required  />
              </label>
              <br/>
              <br/>
              <label>
                Username: <input type="text" name="un" onChange={handleUN}  value={uname}  required />
              </label>
              <br/>
              <br/>
              <label>
                Password: <input type="password" name="passwd" onChange={handlePassw}  value={pword} required />
              </label>
              <br/>
              <br/>
              <div>
                  <input type="submit" value="Submit" />
              </div>
            </form>}
        </div>


    );

}

LocationForm.propTypes ={
  /** Determines whether the form for entering location name, longitude, and latitude values is displayed. */
  display: PropTypes.bool,

  /** Toggles whether the form is displayed. If the remove form is displayed, it will be hidden. */
  toggleDisplay: PropTypes.func,

  /** Called when a location has been successfully added to hide the form. */
  hideDisplay: PropTypes.func,

  /** Called when a location has been successfully added to obtain the updated location list from the server and add the location to the map. */
  obtainData: PropTypes.func,
}
