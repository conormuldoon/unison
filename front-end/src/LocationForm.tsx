import PropTypes from 'prop-types';
import React, { useState } from 'react';
import './App.css';

import { problemConnecting, csrfToken } from './Util';
import HttpStatus from 'http-status-codes';
import { locationPutObject } from './Util';

import parser from 'uri-template';

export function createLocationFactory(obtainData: () => void, selfRef: string, containsRef: string) {

  return function locationFactory(toggleDisplayAdd: () => void, displayAdd: boolean,
    hideAdd: () => void): JSX.Element {

    return <LocationForm obtainData={obtainData} toggleDisplay={toggleDisplayAdd}
      display={displayAdd} hideDisplay={hideAdd} selfRef={selfRef} containsRef={containsRef} />;
  }
}

export interface LocationProps {

  /**
   * A flag that determines whether the add location popup is displayed.
   */
  display: boolean;

  /**
   * A callback invoked to hide the add locatoin popup.
   */
  hideDisplay: () => void;

  /**
   * Toggles whether the add location popup is displayed.
   */
  toggleDisplay: () => void;

  /**
   * A callback for obtaining data from the server that is invoked once a location has been added.
   */
  obtainData: () => void;

  /**
   * A URI for new location PUT requests.
   */
  selfRef: string;

  /**
   * A URI for determining whether the location collection contains a location of a given name.
   */
  containsRef: string;
}

/**
 * A component for displaying a form to add new locations to be tracked.
 * 
 * @component 
 */
function LocationForm({ display, hideDisplay, toggleDisplay, obtainData, selfRef, containsRef }:
  LocationProps): JSX.Element {



  const [location, setLocation] = useState('');
  const [lon, setLon] = useState('');
  const [lat, setLat] = useState('');

  function handleLoc(event: React.ChangeEvent<HTMLInputElement>) {
    setLocation(event.target.value);

  }

  function handleLon(event: React.ChangeEvent<HTMLInputElement>) {
    setLon(event.target.value);
  }

  function handleLat(event: React.ChangeEvent<HTMLInputElement>) {
    setLat(event.target.value);

  }

  function updateDisplay() {
    setLocation('');
    setLon('');
    setLat('');
    hideDisplay();
    obtainData();
  }

  async function putData(selfRef: string) {

    const response = await fetch(selfRef,
      locationPutObject(location, lon, lat));



    if (response.ok) {

      const harvestResponse = await fetch(selfRef + '/' + location, {
        method: 'POST',
        headers: new Headers({
          'X-XSRF-TOKEN': csrfToken()
        }),
      });

      const message = (response.status === HttpStatus.OK) ? 'updated' : 'added';

      if (harvestResponse.ok) {
        alert(location + ' was ' + message + '.');
        updateDisplay();
      } else if (harvestResponse.status === HttpStatus.BAD_GATEWAY) {
        alert(location + ' was ' + message + ', but did did not recieve the weather data. The ' +
          lon + " and " + lat + " longitude and latitude coordinates may not be covered by the model.");
        updateDisplay();
      }
      else {
        alert(location + ' was ' + message + ', but Unison did not obtain the weather data.')
        updateDisplay();
      }
    } else if (response.status === HttpStatus.UNAUTHORIZED) {
      alert('Incorrect user name or password');
    } else if (response.status === HttpStatus.FORBIDDEN) {
      alert('You do not have permission to update ' + location +
        '. A location of the same name may have been added by another user.');
    } else {
      problemConnecting();
    }

  }

  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();

    const template = parser.parse(containsRef);
    const uri = template.expand({ name: location });

    const response = await fetch(uri);

    if (response.ok) {

      const contains = await response.json();
      if (contains.value) {

        if (window.confirm('A location named ' + location + " was previously added. Would you like to update it?")) {
          putData(selfRef);
        }

      } else {
        putData(selfRef);
      }

    } else {
      problemConnecting();
    }


  }


  let text = 'Add Location';
  if (display) {
    text = 'Hide';
  }


  return (<div id='pLeft'>

    <button data-testid="lf-button" onClick={toggleDisplay}>{text}</button>

    {display && <form id='locForm' onSubmit={handleSubmit}>
      <label>
        Location name: <input type="text" name="name" value={location} onChange={handleLoc} required />
      </label>
      <br />
      <br />
      <label>
        Longitude: <input type="number" step="any" name="lon" value={lon} onChange={handleLon} required />
      </label>
      <br />
      <br />
      <label>
        Latitude: <input type="number" step="any" name="lat" value={lat} onChange={handleLat} required />
      </label>
      <br />
      <br />
      <div>
        <input type="submit" value="Submit" />
      </div>
    </form>}
  </div>


  );

}

LocationForm.propTypes = {
  /** Determines whether the form for entering location name, longitude, and latitude values is displayed. */
  display: PropTypes.bool.isRequired,

  /** Toggles whether the form is displayed. If the remove form is displayed, it will be hidden. */
  toggleDisplay: PropTypes.func.isRequired,

  /** Called when a location has been successfully added to hide the form. */
  hideDisplay: PropTypes.func.isRequired,

  /** Called when a location has been successfully added to obtain the updated location list from the server and add the location to the map. */
  obtainData: PropTypes.func.isRequired,

  featureProperties: PropTypes.object,

  selfRef: PropTypes.string,

  containsRef: PropTypes.string,

}

export default LocationForm;
