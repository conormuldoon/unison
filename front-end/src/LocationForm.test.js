import fetchMock from 'fetch-mock';
import React from 'react';
import ReactDOM from 'react-dom';
import { fireEvent, render } from "react-testing-library";
import { API } from './Constant';
import LocationForm from './LocationForm';
import { locationPutObject } from './Util';
import HttpStatus from 'http-status-codes';


it('renders without crashing', async () => {

  const div = document.createElement('div');

  ReactDOM.render(<LocationForm obtainData={() => { }} hideDisplay={() => { }} toggleDisplay={() => { }} display={true} />, div);
  ReactDOM.unmountComponentAtNode(div);
});


it('mathes snapshot', () => {


  const { container } = render(<LocationForm obtainData={() => { }} hideDisplay={() => { }} toggleDisplay={() => { }} display={true} />);

  expect(container).toMatchSnapshot();

});

function changeValue(getByLabelText, labelText, value) {
  fireEvent.change(getByLabelText(labelText), { target: { value: value } });
}

it('posts the data in the form when submit is clicked', async (done) => {
  jest.spyOn(window, 'fetch').mockImplementation(() => { return { headers: { get: () => '0' }, ok: true, json: () => 1 } });
  const hideDisplay = jest.fn();

  const location = 'UCD';

  const lon = '-6.223682';
  const lat = '53.308441';



  const obtainData = () => {


    expect(hideDisplay).toHaveBeenCalledTimes(1);

    expect(fetch).toBeCalledWith(API + '/location', locationPutObject(location, lon, lat));
    done();
  };

  const component = <LocationForm display={true} obtainData={obtainData} hideDisplay={hideDisplay} toggleDisplay={() => { }} />;
  const { getByText, getByLabelText } = render(component);

  changeValue(getByLabelText, 'Location name:', location);
  changeValue(getByLabelText, 'Longitude:', lon);
  changeValue(getByLabelText, 'Latitude:', lat);

  jest.spyOn(window, 'alert').mockImplementation(() => { });

  fireEvent.click(getByText('Submit'));


});

it('handles add location', async (done) => {
  fetchMock.put('end:/location', { status: HttpStatus.OK, body: { json: () => { } } });
  fetchMock.post('end:/harvest', { status: HttpStatus.OK });

  jest.spyOn(window, 'alert').mockImplementation(() => { });
  const hideDisplay = jest.fn();

  const location = 'UCD';
  const obtainData = () => {
    expect(hideDisplay).toHaveBeenCalledTimes(1);
    expect(alert).toBeCalledWith(location + ' was added');
    done();
  };

  const component = <LocationForm display={true} obtainData={obtainData} hideDisplay={hideDisplay} toggleDisplay={() => { }} />;
  const { getByText, getByLabelText } = render(component);

  changeValue(getByLabelText, 'Location name:', location);

  fireEvent.click(getByText('Submit'));

  fetchMock.restore();


});

