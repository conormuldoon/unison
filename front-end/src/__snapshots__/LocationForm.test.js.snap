// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mathes snapshot 1`] = `
<div>
  <div
    id="pLeft"
  >
    <button
      data-testid="lf-button"
    >
      Add Location
    </button>
  </div>
</div>
`;
